### DISCUSSION TITLE
Contrastive motifs from positive and negative series pairs

### DISCUSSION THREAD

**Discussion by hoangthienan95** [TDAmeritrade/stumpy#749] [Q&A]
I'm trying to do the following and wonder if it's possible in STUMPY. I have read the tutorials but they don't seem to quite fit my use case perfectly.

Input: Pairs of series (one of type A and one of type B, not equal length). I have 3 classes of pairs. The distribution of each class is below:
- Class 1:  ~ 600k pairs
- Class 2:    ~ 1k pairs
- Class 3:    ~ 1k pairs

I want to find 
- multi-dimensional motifs (pattern between both type A and B series) that distinguishes each class (either in pairwise or one vs rest fashion).
- multi-dimensional motifs (pattern between both type A and B series) that are similar.

My use case is different from the [Multidimensional_Motif_Discovery tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html) because the motif is repeated in pairs of the same class, not within any series. Also, the motif might not be aligned between type A and B series

For example, the resulting output should be that all Class 1 pairs have the pattern (type A and type B series not matching in sequence but a consistent pair-pattern of length 4)
- Type A series: [2,4,6,8,....]
- Type B series: [2,2,2,2,....]
while pairs in the other classes dont

And all Class 2 pairs have the pattern (type A and type B series matching sequence of length 4 but at different locations )
- Type A series [1,3,5,7,....]
- Type B series [...1,3,5,7....]
while pairs in the other classes dont

I hope I didn't make it too confusing, just trying to see if STUMPY could help with my problem!

**Comment by seanlaw** [comment]
@hoangthienan95 Thank you for your question and welcome to the STUMPY community. So, I fully understand your inputs and outputs but here are a few suggestions:

Since the motifs within your time series pairs are not aligned, you are essentially forced to analyze type A and type B independently for motifs. However, for a given class, I would consider concatenating all type A time series (inserting an `np.nan` between each series so that no motifs will span independent series) and computing the matrix profile via `stumpy.stump(type_A_series, m=4)`. Then, systematically extract the top `n` motifs (maybe `n = 100`). Then, do the same for type B time series. Afterward, do this for the remaining 2 classes. With top `n` motifs for type A and type B in each class, you can then consider doing a pairwise comparison of the motifs across the 3 classes and see if any class has unique motifs that are not found in the others. 

Again, I'm sure I'm not understanding your problem correctly but I believe this should be helpful.
